lib/shopify_api/api_version.rb

```diff
@@ -3,111 +3,143 @@ module ShopifyAPI
   class ApiVersion
     class UnknownVersion < StandardError; end
     class InvalidVersion < StandardError; end
+    include Comparable
 
-    extend DefinedVersions
+    HANDLE_FORMAT = /^\d{4}-\d{2}$/.freeze
+    UNSTABLE_HANDLE = 'unstable'
+    UNSTABLE_AS_DATE = Time.utc(3000, 1, 1)
+    API_PREFIX = '/admin/api/'
+    COERSION_MODES = [:predefined_only, :define_on_unknown].freeze
 
-    include Comparable
+    class << self
+      attr_reader :versions
 
-    def self.coerce_to_version(version_or_name)
-      return version_or_name if version_or_name.is_a?(ApiVersion)
+      def coercion_mode
+        @coercion_mode ||= :define_on_unknown
+      end
 
-      @versions ||= {}
-      @versions.fetch(version_or_name.to_s) do
-        raise UnknownVersion, "#{version_or_name} is not in the defined version set: #{@versions.keys.join(', ')}"
+      def coercion_mode=(mode)
+        raise ArgumentError, "Mode must be one of #{COERSION_MODES}" unless COERSION_MODES.include?(mode)
+        sanitize_known_versions if mode == :predefined_only
+        @coercion_mode = mode
+      end
+
+      def coerce_to_version(version_or_handle)
+        return version_or_handle if version_or_handle.is_a?(ApiVersion)
+        handle = version_or_handle.to_s
+
+        @versions ||= {}
+        @versions.fetch(handle) do
+          if @coercion_mode == :predefined_only
+            error_msg = if @versions.empty?
+              "No versions defined. You must call `ApiVersion.define_known_versions` first."
+            else
+              "`#{handle}` is not in the defined version set. Available versions: #{@versions.keys}"
+            end
+            raise UnknownVersion, "ApiVersion.coercion_mode is set to `:predefined_only`. #{error_msg}"
+          else
+            @versions[handle] = ApiVersion.new(handle: handle)
+          end
+        end
       end
-    end
 
-    def self.define_version(version)
-      @versions ||= {}
+      def define_known_versions
```

Comment:
I might call this `fetch_known_versions` to be explicit that it involves a remote lookup

Reply:
Yeah I think I had that at first and at some point changed it back to minimize the number of changes or something? 🤔 It sorta made sense at the time, but I think you're right that fetch is a better choice.